File: sample/ts/05_mandelbrot_zoom.ts
import { perf_counter } from "./pytra/std/time.js";
import { grayscale_palette } from "./pytra/runtime/gif.js";
import { save_gif } from "./pytra/runtime/gif.js";

// 05: Sample that outputs a Mandelbrot zoom as an animated GIF.

function render_frame(width, height, center_x, center_y, scale, max_iter) {
    let frame = (typeof (width * height) === "number" ? new Array(Math.max(0, Math.trunc(Number((width * height))))).fill(0) : (Array.isArray((width * height)) ? (width * height).slice() : Array.from((width * height))));
    let __hoisted_cast_1 = Number(max_iter);
    const __start_1 = 0;
    for (let y = __start_1; y < height; y += 1) {
        let row_base = y * width;
        let cy = center_y + (y - height * 0.5) * scale;
        const __start_2 = 0;
        for (let x = __start_2; x < width; x += 1) {
            let cx = center_x + (x - width * 0.5) * scale;
            let zx = 0.0;
            let zy = 0.0;
            let i = 0;
            while (i < max_iter) {
                let zx2 = zx * zx;
                let zy2 = zy * zy;
                if (zx2 + zy2 > 4.0) {
                    break;
                }
                zy = 2.0 * zx * zy + cy;
                zx = zx2 - zy2 + cx;
                i += 1;
            }
            frame[(((row_base + x) < 0) ? ((frame).length + (row_base + x)) : (row_base + x))] = Math.trunc(Number(255.0 * i / __hoisted_cast_1));
        }
    }
    return (Array.isArray((frame)) ? (frame).slice() : Array.from((frame)));
}

function run_05_mandelbrot_zoom() {
    let width = 320;
    let height = 240;
    let frame_count = 48;
    let max_iter = 110;
    let center_x = -0.743643887037151;
    let center_y = 0.13182590420533;
    let base_scale = 3.2 / width;
    let zoom_per_frame = 0.93;
    let out_path = "sample/out/05_mandelbrot_zoom.gif";
    
    let start = perf_counter();
    let frames = [];
    let scale = base_scale;
    const __start_3 = 0;
    for (let _ = __start_3; _ < frame_count; _ += 1) {
        frames.push(render_frame(width, height, center_x, center_y, scale, max_iter));
        scale *= zoom_per_frame;
    }
    save_gif(out_path, width, height, frames, grayscale_palette());
    let elapsed = perf_counter() - start;
    console.log("output:", out_path);
    console.log("frames:", frame_count);
    console.log("elapsed_sec:", elapsed);
}

run_05_mandelbrot_zoom();

File: sample/cs/05_mandelbrot_zoom.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Pytra.CsModule;

public static class Program
{
    // 05: Sample that outputs a Mandelbrot zoom as an animated GIF.
    
    public static List<byte> render_frame(long width, long height, double center_x, double center_y, double scale, long max_iter)
    {
        List<byte> frame = Pytra.CsModule.py_runtime.py_bytearray(width * height);
        double __hoisted_cast_1 = System.Convert.ToDouble(max_iter);
        long y = 0;
        for (y = 0; y < height; y += 1) {
            long row_base = y * width;
            double cy = center_y + (y - height * 0.5) * scale;
            long x = 0;
            for (x = 0; x < width; x += 1) {
                double cx = center_x + (x - width * 0.5) * scale;
                double zx = 0.0;
                double zy = 0.0;
                long i = 0;
                while (i < max_iter) {
                    double zx2 = zx * zx;
                    double zy2 = zy * zy;
                    if (zx2 + zy2 > 4.0) {
                        break;
                    }
                    zy = 2.0 * zx * zy + cy;
                    zx = zx2 - zy2 + cx;
                    i += 1;
                }
                Pytra.CsModule.py_runtime.py_set(frame, row_base + x, Pytra.CsModule.py_runtime.py_int(255.0 * i / __hoisted_cast_1));
            }
        }
        return Pytra.CsModule.py_runtime.py_bytes(frame);
    }
    
    public static void run_05_mandelbrot_zoom()
    {
        long width = 320;
        long height = 240;
        long frame_count = 48;
        long max_iter = 110;
        double center_x = -0.743643887037151;
        double center_y = 0.13182590420533;
        double base_scale = 3.2 / width;
        double zoom_per_frame = 0.93;
        string out_path = "sample/out/05_mandelbrot_zoom.gif";
        
        double start = Pytra.CsModule.time.perf_counter();
        System.Collections.Generic.List<List<byte>> frames = new System.Collections.Generic.List<List<byte>>();
        double scale = base_scale;
        long _ = 0;
        for (_ = 0; _ < frame_count; _ += 1) {
            frames.Add(render_frame(width, height, center_x, center_y, scale, max_iter));
            scale *= zoom_per_frame;
        }
        Pytra.CsModule.gif_helper.save_gif(out_path, width, height, frames, Pytra.CsModule.gif_helper.grayscale_palette());
        double elapsed = Pytra.CsModule.time.perf_counter() - start;
        System.Console.WriteLine(string.Join(" ", new object[] { "output:", out_path }));
        System.Console.WriteLine(string.Join(" ", new object[] { "frames:", frame_count }));
        System.Console.WriteLine(string.Join(" ", new object[] { "elapsed_sec:", elapsed }));
    }
    
    public static void Main(string[] args)
    {
            run_05_mandelbrot_zoom();
    }
}
